### PR TITLE
Express the ghost exchange of DoF indices using GridTools::exchange_cell_data_to_ghosts().

### PR DESCRIPTION
**Pull request by bangerth**
This is a very satisfying patch because it removes 325 lines of p4est-heavy code by 125 lines of clean lambda functions that are then passed to the new function @tjhei wrote yesterday in #4788. It is probably easiest to view side-by-side because it's not a change but a wholesale replacement of stuff.

In reference to #3511 and #1366. Also #4704.

### CODE REVIEW

File: source/dofs/dof_handler_policy.cc [hunk 1]
```diff
@@ -3308,191 +3146,166 @@ namespace internal
...
-                          std::find (local_dof_indices.begin(),
-                                     local_dof_indices.end(),
-                                     numbers::invalid_dof_index))
-                        cell->clear_user_flag();
-                    }
-                }
-          }
+            // first see whether we need to do anything at all on this cell.
+            // this is determined by whether the user_flag is set on the
+            // cell that indicates whether the *complete* set of DoF indices
```
Comment (tjhei) [inline source/dofs/dof_handler_policy.cc]: I would say the flag indicates "that the complete set has not been sent"

File: source/dofs/dof_handler_policy.cc [hunk 2]
```diff
@@ -3308,191 +3146,166 @@ namespace internal
...
-              // in the buffer
-              for (unsigned int c=0; c<cells; ++c, dofs+=1+dofs[0])
-                {
-                  const typename DoFHandlerType::level_cell_iterator
-                  cell (&dof_handler.get_triangulation(),
-                        0,
-                        p4est_tree_to_coarse_cell_permutation[cell_data_transfer_buffer.tree_indices[c]],
-                        &dof_handler);
+                Assert (std::find (local_dof_indices.begin(),
+                local_dof_indices.end(),
```
Comment (tjhei) [inline source/dofs/dof_handler_policy.cc]: can you put this into fewer lines, astyle is quite ugly here

**Comment by tjhei** [comment]
Also, please confirm that all MPI tests work (because we don't run those on the integration tester).

**Comment by bangerth** [comment]
As discussed across the table, I'll take care of the two comments in follow-up patches that I already have on top of this one. I verified that all of the MPI tests work.

**Comment by bangerth** [comment]
I rebased this and pushed it again. Let us see what the tester says now. This PR is kinda burning on my nails as it blocks my next few patches.

**Comment by bangerth** [comment]
The one failing test, `lac/tensor_product_matrix_03.debug` would seem to be unrelated. All other tests succeed.

**Comment by tjhei** [comment]
I will keep my fingers crossed that this scales well in parallel.

**Pull request merged**